include/envoy/grpc/async_client_manager.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "envoy/grpc/async_client.h"
+#include "envoy/stats/stats.h"
+
+#include "api/grpc_service.pb.h"
+
+namespace Envoy {
+namespace Grpc {
+
+// Per-service factory for Grpc::AsyncClients. This factory is thread aware and will instantiate
+// with thread local state. Clients will use ThreadLocal::Instance::dispatcher() for event handling.
+class AsyncClientFactory {
+public:
+  virtual ~AsyncClientFactory() {}
+
+  /**
+   * Create a gRPC::AsyncClient.
+   * @return AsyncClientPtr async client.
+   */
+  virtual AsyncClientPtr create() PURE;
+};
+
+typedef std::unique_ptr<AsyncClientFactory> AsyncClientFactoryPtr;
+
+// Singleton gRPC client manager. Grpc::AsyncClientManager can be used to create per-service
+// Grpc::AsyncClientFactory instances.
+class AsyncClientManager {
+public:
+  virtual ~AsyncClientManager() {}
+
+  /**
+   * Create a Grpc::AsyncClients factory for a service. Validation of the service is performed and
+   * will raise an exception on failure.
+   * @param grpc_service envoy::api::v2::GrpcService configuration.
+   * @param scope stats scope.
+   * @return AsyncClientFactoryPtr factory for grpc_service.
+   * @throws EnvoyException when grpc_service validation fails.
+   */
+  virtual AsyncClientFactoryPtr
```

Comment:
I'm not really suggesting we do this in this PR, but I think there is an opportunity here that would further massively cleanup the amount of code that is needed at many of the call sites. Basically, like HTTP async client, if a gRPC client was internally TLS aware, I think many callers would not need any TLS at all. They could just acquire a handle, and then create a TLS local request. I'm pretty sure that if it were implemented this way, TLS in rate limit, access log, and metrics (and soon tracing) would not be needed.

At the very least, I might add some TODOs in this area, or if there any interface changes that would make this easier later we might want to consider it. WDYT?

Reply:
Yeah, as documented, when you `create()` on the factory, the client you get back (like with CM HTTP async client) is the one for your thread. It may be a cheap operation, if all we are returning is a thin handle (not the current implementation, but maybe in the future).

For the Envoy gRPC client today, `Grpc::AsyncClientImpl` uses the HTTP async client from CM, so it already is somewhat TLS aware. In the Google gRPC world, we might have things like TLS completion queues that are managed by the the TLS for the Google gRPC client.

The question is then, at these various sites that use the gRPC async client, do they have any specific TLS state that is external to the client. Looking at access logs gRPC, there is the `stream_map_`; would this remain TLS or become some state that is shared across threads? There's a tradeoff here; introducing more complexity (or contention) on shared structures vs. more streams (one per silo) for the same resource.

Reply:
> The question is then, at these various sites that use the gRPC async client, do they have any specific TLS state that is external to the client. Looking at access logs gRPC, there is the stream_map_; would this remain TLS or become some state that is shared across threads? There's a tradeoff here; introducing more complexity (or contention) on shared structures vs. more streams (one per silo) for the same resource.

Yeah I'm not sure. The main pattern that I think would be possible if the client itself was TLS aware is you could grab a client and do error checking, and then safely use that client across all threads to start streams. In this case, the client itself is TLS aware, but it still effectively needs to be stored in a local TLS slot for use.

It's not a big deal, this is fine as is. We can look at further cleanups later.